src/parser/ts/parser.unit.test.ts

```diff
@@ -0,0 +1,443 @@
+  ];
+  inputs.map(input => it(input, () => {
+    expect(reverse(parse(input))).toBe(input);
+  }));
+})
+
+describe('inject ids in queries', () => {
+  it('function with arithmetics', () => {
+    const func = STSFunction('sin', [
+      STS([Filter('assetSubtreeIds', [Filter('id', 1)])]),
+      Operator('/'),
+      Constant(10),
+    ]);
+    const res = injectTSIdsInExpression(func, [[{ id: 1 } as TimeSeriesResponseItem]]);
+    expect(res).toEqual('sin(ts{id=1} / 10)');
+  });
+
+  it('multiary function', () => {
+    const func = STSFunction('avg', [
+      STS([Filter('assetSubtreeIds', [Filter('id', 1)])]),
+    ]);
+    const res = injectTSIdsInExpression(func, [
+      [{ id: 1 }, { id: 2 }, { id: 3 }],
+    ] as TimeSeriesResponseItem[][]);
+    expect(res).toEqual('avg(ts{id=1}, ts{id=2}, ts{id=3})');
+  });
+
+  it('sum function with one filter', () => {
+    const func = STSFunction('sum', [
+      STS([Filter('assetSubtreeIds', [Filter('id', 1)])]),
+    ]);
+    const res = injectTSIdsInExpression(func, [
+      [{ id: 1 }, { id: 2 }, { id: 3 }],
+    ] as TimeSeriesResponseItem[][]);
+    expect(res).toEqual('(ts{id=1} + ts{id=2} + ts{id=3})');
+  });
+});
+
+describe('find multiary function indices', () => {
+  const funcWithOneParam = STSFunction('avg', [STS()]);
+  const funcWithFewParams = STSFunction('avg', [STS(), STS()]);
+  const funcComposition = [
+    funcWithFewParams,
+    Operator('+'),
+    funcWithOneParam,
+    Operator('/'),
+    funcWithOneParam,
+  ];
+  const nestedComposition = [
+    STSFunction('sin', funcComposition),
+    Operator('+'),
+    STSFunction('avg', funcComposition),
+  ];
+  const indices = getIndicesOfMultiaryFunctionArgs;
+
+  it('valid for single param', () => {
+    const res = indices(cloneDeep(funcWithOneParam));
+    expect(res).toEqual([0]);
+  });
+
+  it('invalid if multiple params', () => {
+    const res = indices(cloneDeep(funcWithFewParams));
+    expect(res).toEqual([]);
+  });
+
+  it('works with composition of functions', () => {
+    const res = indices(cloneDeep(funcComposition));
+    expect(res).toEqual([2, 3]);
+  });
+
+  it('works with nested composition of functions', () => {
+    const res = indices(cloneDeep(STSFunction('sin', funcComposition)));
+    expect(res).toEqual([2, 3]);
+  });
+
+  it('works with complex nested composition of functions', () => {
+    const res = indices(cloneDeep(nestedComposition));
+    expect(res).toEqual([2, 3, 6, 7]);
+  });
+
+  it('map() is a special case, doesn\'t work as multiary', () => {
+    const mapFunc = STSFunction('map', [STS(), ['1'], [1], 1]);
+    const res = indices(mapFunc);
+    expect(res).toEqual([]);
+  });
+});
+
+describe('permutations', () => {
+  const gen = generateAllPossiblePermutations;
+
+  it('one element', () => {
+    expect(gen([['a']])).toEqual([[['a']]]);
+  });
+
+  it('multiple elements', () => {
+    expect(gen([['a'], ['1', '2', '3'], ['å', 'æ']])).toEqual([
```

Comment:
Nice) Now I at least understand what `generateAllPossiblePermutations` actually do 😅

Reply:
yeah I know 😅
but at least it's abstracted a bit from other logic